File: jas_ocaml/test/tools/yaml_tool_hit_test_test.ml
(** Magic Wand + Eyedropper gesture-seam tests — OCaml port of the Rust
    seam tests in jas_dioxus/src/tools/yaml_tool.rs (the
    magic_wand_parity_* and eyedropper_parity_* families). Drives the
    PRODUCTION magic_wand / eyedropper tools loaded from the workspace
    bundle through on_press / on_release and asserts the live selection
    paths and the sampled fill colour, matching the Rust reference numbers
    and tolerances exactly.

    Both tools fire on on_mousedown: each does hit_test(event.x, event.y)
    on the live registered document (the same headless path the selection
    tools use — the always-miss hit-test stubs in make_ctx are inert here),
    then dispatches doc.magic_wand.apply or doc.eyedropper.sample /
    apply_loaded. The seam runs the full pipeline (snapshot, hit-test,
    branch, effect) rather than calling the effect directly.

    Magic Wand config seeding. The wand commit reads the nine app-level
    state.magic_wand_* options. Those keys are bridged into the tool global
    namespace by the PRODUCTION bridge (bridge_app_state, the OCaml analog
    of the Rust sync_global_state) — and ONLY because the nine keys were
    added to bridged_state_keys (yaml_tool.ml) alongside this test. Before
    that fix the live wand always fell back to Magic_wand.default_config and
    SILENTLY IGNORED every Magic Wand Panel adjustment. The
    respects_bridged_nondefault_config case is the regression gate: it
    routes magic_wand_fill_color=false through bridge_app_state and proves
    it changes the wand result. Remove the keys from the allowlist and that
    case fails.

    Eyedropper needs no bridge seeding: the state.eyedropper_* toggles all
    default true and Eyedropper.default_config agrees, so the fill-copy
    path falls back to all-on and the cache write goes straight to the tool
    store. *)

open Jas

let () = ignore (GMain.init ())

(* ── Loaders ───────────────────────────────────────── *)

let load_tool (key : string) : Yaml_tool.yaml_tool option =
  match Workspace_loader.load () with
  | None -> None
  | Some ws ->
    match Workspace_loader.json_member "tools" ws.data with
    | Some (`Assoc tools) ->
      (match List.assoc_opt key tools with
       | Some spec -> Yaml_tool.from_workspace_tool spec
       | None -> None)
    | _ -> None

let magic_wand_tool () = load_tool "magic_wand"
let eyedropper_tool () = load_tool "eyedropper"

(* ── Shared helpers ────────────────────────────────── *)

let make_ctx model =
  let ctrl = new Controller.controller ~model () in
  let ctx : Canvas_tool.tool_context = {
    model;
    controller = ctrl;
    hit_test_selection = (fun _ _ -> false);
    hit_test_handle = (fun _ _ -> None);
    hit_test_text = (fun _ _ -> None);
    hit_test_path_curve = (fun _ _ -> None);
    request_update = (fun () -> ());
    draw_element_overlay = (fun _cr _elem ~is_partial:_ _cps -> ());
  } in
  (ctx, ctrl)

(** A 1pt black stroke, identical across every fixture rect. Mirrors the
    Rust Stroke::new(Color::BLACK, 1.0). *)
let black_stroke : Element.stroke = {
  stroke_color = Element.black;
  stroke_width = 1.0;
  stroke_linecap = Element.Butt;
  stroke_linejoin = Element.Miter;
  stroke_miter_limit = 4.0;
  stroke_align = Element.Center;
  stroke_dash_pattern = [];
  stroke_dash_align_anchors = false;
  stroke_start_arrow = Element.Arrow_none;
  stroke_end_arrow = Element.Arrow_none;
  stroke_start_arrow_scale = 1.0;
  stroke_end_arrow_scale = 1.0;
  stroke_arrow_align = Element.Tip_at_end;
  stroke_opacity = 1.0;
}

let make_rect ?fill (x : float) : Element.element =
  Element.Rect { name = None; id = None;
    x; y = 0.0; width = 10.0; height = 10.0; rx = 0.0; ry = 0.0;
    fill; stroke = Some black_stroke; opacity = 1.0;
    transform = None; locked = false; visibility = Element.Preview;
    blend_mode = Element.Normal; mask = None;
    fill_gradient = None; stroke_gradient = None;
  }

let rect_fill = function
  | Element.Rect { fill; _ } -> fill
  | _ -> assert false

(** Selected element paths as a set, for order-independent assertions.
    Mirrors the Rust selection_paths. *)
let selection_paths (m : Model.model) : (int list) list =
  Document.PathMap.bindings m#document.selection
  |> List.map (fun (_, (es : Document.element_selection)) -> es.es_path)
  |> List.sort compare

(** Replace the selection with exactly [paths] (each whole-element),
    mirroring the Rust Controller::set_selection with ElementSelection::all. *)
let set_selection (ctrl : Controller.controller) (paths : int list list) : unit =
  let sel = List.fold_left (fun acc p ->
    Document.PathMap.add p (Document.element_selection_all p) acc
  ) Document.PathMap.empty paths in
  ctrl#set_selection sel

(* ── Magic Wand fixture ────────────────────────────── *)

(** Three rects in one layer — red @[0,0] (x=0), red @[0,1] (x=20),
    blue @[0,2] (x=40) — each 10x10 with an identical 1pt black stroke
    and opacity 1.0. Mirrors the Rust magic_wand_seam_model. *)
let magic_wand_seam_model () : Model.model =
  let red = Element.make_fill (Element.color_rgb 1.0 0.0 0.0) in
  let blue = Element.make_fill (Element.color_rgb 0.0 0.0 1.0) in
  let layer = Element.Layer {
    name = Some "L"; id = None;
    children = [|
      make_rect ~fill:red 0.0;
      make_rect ~fill:red 20.0;
      make_rect ~fill:blue 40.0;
    |];
    transform = None; locked = false; opacity = 1.0;
    visibility = Element.Preview; blend_mode = Element.Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

(** The full default Magic Wand config as the workspace ships it, routed
    through the PRODUCTION bridge (bridge_app_state ~overrides, the OCaml
    analog of the Rust sync_global_state) so the tool store carries it
    exactly as the live canvas would. Mirrors the Rust
    seed_magic_wand_defaults. *)
let seed_magic_wand_defaults (tool : Yaml_tool.yaml_tool)
    (model : Model.model) : unit =
  tool#bridge_app_state ~overrides:[
    ("magic_wand_fill_color", `Bool true);
    ("magic_wand_fill_tolerance", `Int 32);
    ("magic_wand_stroke_color", `Bool true);
    ("magic_wand_stroke_tolerance", `Int 32);
    ("magic_wand_stroke_weight", `Bool true);
    ("magic_wand_stroke_weight_tolerance", `Float 5.0);
    ("magic_wand_opacity", `Bool true);
    ("magic_wand_opacity_tolerance", `Int 5);
    ("magic_wand_blending_mode", `Bool false);
  ] model

(* ── Magic Wand tests ──────────────────────────────── *)

let magic_wand_tests = [
  Alcotest.test_case "click_red_selects_both_reds_not_blue" `Quick (fun () ->
    match magic_wand_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = magic_wand_seam_model () in
      let (ctx, _) = make_ctx m in
      seed_magic_wand_defaults tool m;
      (* Plain click on the first red rect at screen (5,5) -> replace. *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:false;
      let paths = selection_paths m in
      assert (List.mem [0; 0] paths);
      assert (List.mem [0; 1] paths);
      assert (not (List.mem [0; 2] paths));
      assert (List.length paths = 2));

  Alcotest.test_case "click_blue_selects_only_blue" `Quick (fun () ->
    match magic_wand_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = magic_wand_seam_model () in
      let (ctx, _) = make_ctx m in
      seed_magic_wand_defaults tool m;
      (* Plain click on the blue rect at screen (45,5) -> replace. *)
      tool#on_press ctx 45.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 45.0 5.0 ~shift:false ~alt:false;
      assert (selection_paths m = [[0; 2]]));

  Alcotest.test_case "shift_click_unions_alt_click_subtracts" `Quick (fun () ->
    match magic_wand_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = magic_wand_seam_model () in
      let (ctx, ctrl) = make_ctx m in
      seed_magic_wand_defaults tool m;
      (* Pre-select the blue rect [0,2]. *)
      set_selection ctrl [[0; 2]];
      (* Shift+click red [0,0] -> ADD: {2} union {0,1} = {0,1,2}. *)
      tool#on_press ctx 5.0 5.0 ~shift:true ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:true ~alt:false;
      assert (selection_paths m = [[0; 0]; [0; 1]; [0; 2]]);
      (* Alt+click red [0,0] -> SUBTRACT the wand result {0,1}: leaves {2}. *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:true;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:true;
      assert (selection_paths m = [[0; 2]]));

  Alcotest.test_case "click_empty_clears_selection" `Quick (fun () ->
    match magic_wand_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = magic_wand_seam_model () in
      let (ctx, ctrl) = make_ctx m in
      seed_magic_wand_defaults tool m;
      (* Start with a non-empty selection. *)
      set_selection ctrl [[0; 1]];
      assert (not (Document.PathMap.is_empty m#document.selection));
      (* Plain click on empty canvas (100,100) -> selection cleared. *)
      tool#on_press ctx 100.0 100.0 ~shift:false ~alt:false;
      tool#on_release ctx 100.0 100.0 ~shift:false ~alt:false;
      assert (Document.PathMap.is_empty m#document.selection));

  Alcotest.test_case "respects_bridged_nondefault_config" `Quick (fun () ->
    (* REGRESSION GATE for the live state-bridge fix. With Fill Color
       turned OFF and only stroke / weight / opacity matching the seed, the
       blue rect — which has the SAME 1pt black stroke and opacity as the
       reds — now also matches, so a click on a red selects ALL THREE rects.
       This non-default config reaches the tool only via bridge_app_state,
       and only because magic_wand_* is now in bridged_state_keys. Remove
       the keys from the allowlist and the config falls back to
       Magic_wand.default_config (Fill ON) -> the blue stops matching ->
       this assertion fails. That is the bridge proof. *)
    match magic_wand_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = magic_wand_seam_model () in
      let (ctx, _) = make_ctx m in
      tool#bridge_app_state ~overrides:[
        ("magic_wand_fill_color", `Bool false);
        ("magic_wand_fill_tolerance", `Int 32);
        ("magic_wand_stroke_color", `Bool true);
        ("magic_wand_stroke_tolerance", `Int 32);
        ("magic_wand_stroke_weight", `Bool true);
        ("magic_wand_stroke_weight_tolerance", `Float 5.0);
        ("magic_wand_opacity", `Bool true);
        ("magic_wand_opacity_tolerance", `Int 5);
        ("magic_wand_blending_mode", `Bool false);
      ] m;
      (* Click red [0,0]. Fill is ignored, stroke + weight + opacity are
         identical across all three rects -> all three match. *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:false;
      assert (selection_paths m = [[0; 0]; [0; 1]; [0; 2]]));
]

(* ── Eyedropper fixture ────────────────────────────── *)

(** The exact green the eyedropper fixture source carries — a distinctive
    non-primary colour so the apply assertion cannot accidentally pass
    against a stray black / red default. Mirrors the Rust
    eyedropper_source_color rgb(0.0, 0.6, 0.2). *)
let eyedropper_source_color : Element.color =
  Element.color_rgb 0.0 0.6 0.2

(** Two rects in one layer: source [0,0] green-filled (x=0), target [0,1]
    fill-less (x=20). Both 10x10 at the identity view, so screen (5,5) hits
    the source and screen (25,5) hits the target. Mirrors the Rust
    eyedropper_seam_model. *)
let eyedropper_seam_model () : Model.model =
  let green = Element.make_fill eyedropper_source_color in
  let layer = Element.Layer {
    name = Some "L"; id = None;
    children = [|
      make_rect ~fill:green 0.0;
      make_rect 20.0;
    |];
    transform = None; locked = false; opacity = 1.0;
    visibility = Element.Preview; blend_mode = Element.Normal; mask = None;
    isolated_blending = false; knockout_group = false;
  } in
  let m = Model.create () in
  m#set_document_unbracketed (Document.make_document [| layer |]);
  m

(** Assert two colours are equal channel-for-channel in RGBA, mirroring
    the exact-equality Rust assert on fill.color. *)
let assert_color_eq (expected : Element.color) (actual : Element.color)
    (msg : string) : unit =
  let (er, eg, eb, ea) = Element.color_to_rgba expected in
  let (ar, ag, ab, aa) = Element.color_to_rgba actual in
  if not (er = ar && eg = ag && eb = ab && ea = aa) then begin
    Printf.eprintf "%s: expected (%f,%f,%f,%f) got (%f,%f,%f,%f)\n%!"
      msg er eg eb ea ar ag ab aa;
    assert false
  end

(* ── Eyedropper tests ──────────────────────────────── *)

let eyedropper_tests = [
  Alcotest.test_case "click_source_with_selection_copies_fill_to_target"
    `Quick (fun () ->
    match eyedropper_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = eyedropper_seam_model () in
      let (ctx, ctrl) = make_ctx m in
      (* Pre-select the empty target [0,1]; the source [0,0] is clicked. *)
      set_selection ctrl [[0; 1]];
      assert (rect_fill (Document.get_element m#document [0; 1]) = None);
      (* Plain click on the green source at screen (5,5) -> sample, which
         (selection non-empty) also writes the appearance to [0,1]. *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:false;
      (match rect_fill (Document.get_element m#document [0; 1]) with
       | None -> assert false
       | Some f ->
         assert_color_eq eyedropper_source_color f.fill_color
           "eyedropper sample must copy the exact source green into target"));

  Alcotest.test_case "alt_click_applies_cached_color_to_target"
    `Quick (fun () ->
    match eyedropper_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = eyedropper_seam_model () in
      let (ctx, _) = make_ctx m in
      (* First, plain-click the source with NO selection -> loads the cache
         (and mutates nothing, since the selection is empty). *)
      tool#on_press ctx 5.0 5.0 ~shift:false ~alt:false;
      tool#on_release ctx 5.0 5.0 ~shift:false ~alt:false;
      assert (rect_fill (Document.get_element m#document [0; 1]) = None);
      (* Now Alt+click the empty target [0,1] at screen (25,5) ->
         apply_loaded writes the cached green into the target. *)
      tool#on_press ctx 25.0 5.0 ~shift:false ~alt:true;
      tool#on_release ctx 25.0 5.0 ~shift:false ~alt:true;
      (match rect_fill (Document.get_element m#document [0; 1]) with
       | None -> assert false
       | Some f ->
         assert_color_eq eyedropper_source_color f.fill_color
           "Alt+click must apply the cached green to the target"));

  Alcotest.test_case "click_empty_is_a_noop" `Quick (fun () ->
    match eyedropper_tool () with
    | None -> Alcotest.skip ()
    | Some tool ->
      let m = eyedropper_seam_model () in
      let (ctx, _) = make_ctx m in
      (* Snapshot the document layers before the gesture for an exact
         no-op proof. *)
      let before = m#document.layers in
      (* Plain click on empty canvas (100,100) -> no hit -> no-op. *)
      tool#on_press ctx 100.0 100.0 ~shift:false ~alt:false;
      tool#on_release ctx 100.0 100.0 ~shift:false ~alt:false;
      assert (m#document.layers = before);
      (* The source fill is untouched; the target is still fill-less. *)
      (match rect_fill (Document.get_element m#document [0; 0]) with
       | Some f -> assert_color_eq eyedropper_source_color f.fill_color
                     "source fill untouched after empty-space click"
       | None -> assert false);
      assert (rect_fill (Document.get_element m#document [0; 1]) = None));
]

let () =
  Alcotest.run "Yaml hit-test tools" [
    "Magic Wand", magic_wand_tests;
    "Eyedropper", eyedropper_tests;
  ]

File: jas/tools/yaml_tool_hit_test_test.py
"""Hit-test gesture-seam tests for the Magic Wand and Eyedropper YAML tools.

Ports the magic_wand_parity_* (5) and eyedropper_parity_* (3) seam tests
from the Rust reference (jas_dioxus/src/tools/yaml_tool.rs) 1:1. Both tools
resolve a clicked element with ``hit_test(event.x, event.y)`` — which
already runs headlessly in the seam (the same primitive the selection
tools rely on, registered per-dispatch by ``YamlTool._dispatch`` via
``doc_primitives.register_document``) — so no new hit-test fixture is
needed; the cases reuse the loader + ctx machinery from
``yaml_tool_selection_variants_test.py``.

Two tools, eight cases:

  * magic_wand — click a red rect selects BOTH reds (not blue); click
    blue selects only blue; shift-click unions, alt-click subtracts;
    click empty clears; and a non-default-config gate that seeds
    ``magic_wand_fill_color=false`` through the PRODUCTION app-state
    bridge (``seed_globals_from`` / the ``BRIDGED_STATE_KEYS`` allowlist,
    the same path the live Magic Wand Panel drives) so the blue — sharing
    the reds' 1pt black stroke and opacity — also matches and a click on
    a red selects all three rects.
  * eyedropper — click a green source with an empty target selected
    copies the EXACT source green (0,0.6,0.2) into the target; a
    plain-click loads the cache then an alt-click applies it; a click on
    empty space is a byte-identical no-op.

The asserted selection paths, sampled rgb, and config numbers/tolerances
mirror the Rust reference exactly.
"""

from __future__ import annotations

import json
import os
import sys

_REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
if _REPO_ROOT not in sys.path:
    sys.path.insert(0, _REPO_ROOT)
_JAS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if _JAS_DIR not in sys.path:
    sys.path.insert(0, _JAS_DIR)

import pytest

from document.controller import Controller
from document.document import Document, ElementSelection
from document.model import Model
from geometry.element import Color, Fill, Layer, Rect as RectElem, Stroke
from tools.yaml_tool import BRIDGED_STATE_KEYS
from workspace_interpreter.state_store import StateStore


# ── Shared machinery (mirrors yaml_tool_selection_variants_test.py) ──


def _load_ws_tool(tool_id: str) -> "YamlTool | None":
    from tools.yaml_tool import YamlTool

    ws_path = os.path.abspath(os.path.join(
        _REPO_ROOT, "workspace", "workspace.json",
    ))
    if not os.path.exists(ws_path):
        return None
    with open(ws_path, "r") as f:
        data = json.load(f)
    tools = data.get("tools")
    if not isinstance(tools, dict):
        return None
    spec = tools.get(tool_id)
    return YamlTool.from_workspace_tool(spec) if spec else None


def _ctx(model: Model, app_state: StateStore | None = None):
    """Build the ad-hoc ToolContext the YAML tools dispatch against. When
    ``app_state`` is given it is exposed as ``ctx.app_state`` so the
    PRODUCTION per-dispatch bridge (``_store.seed_globals_from`` over
    ``BRIDGED_STATE_KEYS``) copies the seeded ``state.*`` keys into the
    tool store before each handler runs — the same path the live canvas
    uses to feed Magic Wand Panel values to the wand."""
    ctrl = Controller(model)
    ctx_obj = type("Ctx", (), {})()
    ctx_obj.model = model
    ctx_obj.controller = ctrl
    ctx_obj.document = model.document
    ctx_obj.request_update = lambda: None
    ctx_obj.app_state = app_state
    return ctx_obj, ctrl


def _selection_paths(model: Model) -> set[tuple[int, ...]]:
    """Selected element paths as a set, for order-independent assertions
    (mirrors Rust's selection_paths)."""
    return {es.path for es in model.document.selection}


# ── Magic Wand fixtures (mirror magic_wand_seam_model) ──────────────


def _magic_wand_seam_model() -> Model:
    """Three rects in one layer — red @[0,0], red @[0,1], blue @[0,2] —
    each 10x10 with an identical 1pt black stroke, laid out at x = 0,
    20, 40 so screen (5,5) hits the first red, (45,5) hits the blue.
    Mirrors the Rust magic_wand_seam_model fixture (same geometry the
    effect tests assert against)."""
    red = Fill(color=Color.rgb(1.0, 0.0, 0.0))
    blue = Fill(color=Color.rgb(0.0, 0.0, 1.0))
    stroke = Stroke(color=Color.rgb(0.0, 0.0, 0.0), width=1.0)

    def make(fill: Fill, x: float) -> RectElem:
        return RectElem(x=x, y=0.0, width=10.0, height=10.0,
                        fill=fill, stroke=stroke)

    layer = Layer(name="L", children=(
        make(red, 0.0),
        make(red, 20.0),
        make(blue, 40.0),
    ))
    return Model(document=Document(layers=(layer,)))


def _magic_wand_app_state(fill_color: bool = True) -> StateStore:
    """The full Magic Wand config, written to an app-state store so the
    PRODUCTION bridge carries it into the tool store exactly as the live
    canvas would. ``fill_color`` toggles the one criterion the
    non-default-config gate flips. Mirrors Rust's seed_magic_wand_defaults
    / the sync_global_state map."""
    s = StateStore()
    s.set("magic_wand_fill_color", fill_color)
    s.set("magic_wand_fill_tolerance", 32)
    s.set("magic_wand_stroke_color", True)
    s.set("magic_wand_stroke_tolerance", 32)
    s.set("magic_wand_stroke_weight", True)
    s.set("magic_wand_stroke_weight_tolerance", 5.0)
    s.set("magic_wand_opacity", True)
    s.set("magic_wand_opacity_tolerance", 5)
    s.set("magic_wand_blending_mode", False)
    return s


# ── Magic Wand seam tests ───────────────────────────────────────────


class TestMagicWandHitTest:
    def test_click_red_selects_both_reds_not_blue(self):
        tool = _load_ws_tool("magic_wand")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _magic_wand_seam_model()
        ctx, _ = _ctx(model, _magic_wand_app_state())

        # Plain click on the first red rect at screen (5,5) -> replace.
        tool.on_press(ctx, 5.0, 5.0)
        tool.on_release(ctx, 5.0, 5.0)

        paths = _selection_paths(model)
        assert (0, 0) in paths, f"seed red [0,0] selected, got {paths}"
        assert (0, 1) in paths, f"matching red [0,1] selected, got {paths}"
        assert (0, 2) not in paths, f"blue [0,2] must NOT be selected, got {paths}"
        assert paths == {(0, 0), (0, 1)}, f"exactly the two reds, got {paths}"

    def test_click_blue_selects_only_blue(self):
        tool = _load_ws_tool("magic_wand")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _magic_wand_seam_model()
        ctx, _ = _ctx(model, _magic_wand_app_state())

        # Plain click on the blue rect at screen (45,5) -> replace.
        tool.on_press(ctx, 45.0, 5.0)
        tool.on_release(ctx, 45.0, 5.0)

        assert _selection_paths(model) == {(0, 2)}, (
            "clicking blue selects ONLY blue [0,2] (no red matches), "
            f"got {_selection_paths(model)}"
        )

    def test_shift_click_unions_alt_click_subtracts(self):
        tool = _load_ws_tool("magic_wand")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _magic_wand_seam_model()
        ctx, ctrl = _ctx(model, _magic_wand_app_state())

        # Pre-select the blue rect [0,2].
        ctrl.set_selection(frozenset({ElementSelection.all((0, 2))}))

        # Shift+click red [0,0] -> ADD: {2} u {0,1} = {0,1,2}.
        tool.on_press(ctx, 5.0, 5.0, shift=True)
        tool.on_release(ctx, 5.0, 5.0, shift=True)
        assert _selection_paths(model) == {(0, 0), (0, 1), (0, 2)}, (
            "Shift+click unions the wand result onto the existing selection"
        )

        # Alt+click red [0,0] -> SUBTRACT the wand result {0,1}: leaves {2}.
        tool.on_press(ctx, 5.0, 5.0, alt=True)
        tool.on_release(ctx, 5.0, 5.0, alt=True)
        assert _selection_paths(model) == {(0, 2)}, (
            "Alt+click subtracts the wand result {0,1}, leaving blue {2}"
        )

    def test_click_empty_clears_selection(self):
        tool = _load_ws_tool("magic_wand")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _magic_wand_seam_model()
        ctx, ctrl = _ctx(model, _magic_wand_app_state())

        # Start with a non-empty selection.
        ctrl.set_selection(frozenset({ElementSelection.all((0, 1))}))
        assert len(model.document.selection) != 0

        # Plain click on empty canvas (100,100) -> selection cleared.
        tool.on_press(ctx, 100.0, 100.0)
        tool.on_release(ctx, 100.0, 100.0)
        assert len(model.document.selection) == 0, (
            "plain click on empty canvas clears the selection, "
            f"got {_selection_paths(model)}"
        )

    def test_respects_bridged_nondefault_config(self):
        # REGRESSION GATE for the live state-bridge fix. With Fill Color
        # turned OFF and only stroke/weight/opacity matching the seed, the
        # blue rect — which has the SAME 1pt black stroke and opacity as
        # the reds — also matches, so a click on a red selects all THREE
        # rects. This non-default config only reaches the tool via the
        # bridge, and only because magic_wand_* is now in
        # BRIDGED_STATE_KEYS. Drop the keys from the allowlist and the
        # config falls back to MagicWandConfig() (Fill ON) -> the blue
        # stops matching -> this assertion fails. That is the bridge proof.
        tool = _load_ws_tool("magic_wand")
        if tool is None:
            pytest.skip("workspace.json not available")
        # Sanity: the keys this gate depends on are actually bridged.
        assert "magic_wand_fill_color" in BRIDGED_STATE_KEYS
        model = _magic_wand_seam_model()
        ctx, _ = _ctx(model, _magic_wand_app_state(fill_color=False))

        # Click red [0,0]. Fill is ignored, stroke+weight+opacity are
        # identical across all three rects -> all three match.
        tool.on_press(ctx, 5.0, 5.0)
        tool.on_release(ctx, 5.0, 5.0)

        assert _selection_paths(model) == {(0, 0), (0, 1), (0, 2)}, (
            "with Fill Color OFF (bridged), the wand matches on shared "
            "stroke/weight/opacity -> all three rects; got "
            f"{_selection_paths(model)}"
        )


# ── Eyedropper fixtures (mirror eyedropper_seam_model) ──────────────


def _eyedropper_source_color() -> Color:
    """The exact green the eyedropper fixture source carries — a
    distinctive non-primary colour so the apply assertion can't
    accidentally pass against a stray black/red default. (0,0.6,0.2)
    round-trips through the appearance cache exactly: 0.6*255 = 153,
    0.2*255 = 51, both integral."""
    return Color.rgb(0.0, 0.6, 0.2)


def _eyedropper_seam_model() -> Model:
    """Two rects in one layer: source [0,0] green-filled, target [0,1]
    fill-less, both 10x10 side by side at x = 0 and 20 so screen (5,5)
    hits the source and (25,5) hits the target. Mirrors the Rust
    eyedropper_seam_model fixture."""
    green = Fill(color=_eyedropper_source_color())
    stroke = Stroke(color=Color.rgb(0.0, 0.0, 0.0), width=1.0)

    def make(fill: Fill | None, x: float) -> RectElem:
        return RectElem(x=x, y=0.0, width=10.0, height=10.0,
                        fill=fill, stroke=stroke)

    layer = Layer(name="L", children=(
        make(green, 0.0),
        make(None, 20.0),
    ))
    return Model(document=Document(layers=(layer,)))


# ── Eyedropper seam tests ───────────────────────────────────────────
#
# The Eyedropper is a single-click tool: on_mousedown hit-tests and (on a
# hit) fires doc.eyedropper.sample (plain) or doc.eyedropper.apply_loaded
# (Alt). sample snapshots the source appearance into state.eyedropper_cache
# AND — when the selection is non-empty — writes that appearance to every
# eligible selected target. A click on empty space is a no-op. The
# eyedropper toggles all default true and EyedropperConfig() agrees, so the
# fill-copy path needs no bridge seeding (the cache write goes straight to
# the tool store and the config falls back to all-on).


class TestEyedropperHitTest:
    def test_click_source_with_selection_copies_fill_to_target(self):
        tool = _load_ws_tool("eyedropper")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _eyedropper_seam_model()
        ctx, ctrl = _ctx(model)

        # Pre-select the empty target [0,1]; the source [0,0] is clicked.
        ctrl.set_selection(frozenset({ElementSelection.all((0, 1))}))
        assert model.document.get_element((0, 1)).fill is None, (
            "precondition: target starts with no fill"
        )

        # Plain click on the green source at screen (5,5) -> sample, which
        # (selection non-empty) also writes the appearance to [0,1].
        tool.on_press(ctx, 5.0, 5.0)
        tool.on_release(ctx, 5.0, 5.0)

        fill = model.document.get_element((0, 1)).fill
        assert fill is not None, "target now carries the sampled fill"
        assert fill.color == _eyedropper_source_color(), (
            "eyedropper sample must copy the EXACT source green (0,0.6,0.2) "
            f"into the selected target, got {fill.color}"
        )

    def test_alt_click_applies_cached_color_to_target(self):
        tool = _load_ws_tool("eyedropper")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _eyedropper_seam_model()
        ctx, _ = _ctx(model)

        # First, plain-click the source with NO selection -> loads the
        # cache (and mutates nothing, since the selection is empty).
        tool.on_press(ctx, 5.0, 5.0)
        tool.on_release(ctx, 5.0, 5.0)
        assert model.document.get_element((0, 1)).fill is None, (
            "a sample with no selection must not mutate other elements"
        )

        # Now Alt+click the empty target [0,1] at screen (25,5) ->
        # apply_loaded writes the cached green into the target.
        tool.on_press(ctx, 25.0, 5.0, alt=True)
        tool.on_release(ctx, 25.0, 5.0, alt=True)

        fill = model.document.get_element((0, 1)).fill
        assert fill is not None, "apply_loaded wrote the cached fill"
        assert fill.color == _eyedropper_source_color(), (
            "Alt+click must apply the cached green (0,0.6,0.2) to the "
            f"target, got {fill.color}"
        )

    def test_click_empty_is_a_noop(self):
        tool = _load_ws_tool("eyedropper")
        if tool is None:
            pytest.skip("workspace.json not available")
        model = _eyedropper_seam_model()
        ctx, _ = _ctx(model)

        # Snapshot the document before the gesture for an exact-equality
        # no-op proof.
        before = repr(model.document.layers)

        # Plain click on empty canvas (100,100) -> no hit -> no-op.
        tool.on_press(ctx, 100.0, 100.0)
        tool.on_release(ctx, 100.0, 100.0)

        assert repr(model.document.layers) == before, (
            "a click on empty space must not mutate the document"
        )
        # The source fill is untouched; the target is still fill-less.
        assert model.document.get_element((0, 0)).fill.color == \
            _eyedropper_source_color()
        assert model.document.get_element((0, 1)).fill is None, (
            "target remains fill-less after an empty-space click"
        )
